### ISSUE TITLE
Flatpak build

### ISSUE THREAD

**Issue by LeandroStanger** [flutter/flutter#41737]
Please, consider supporting [Flatpak](http://docs.flatpak.org/en/latest/introduction.html#reasons-to-use-flatpak) and publishing flutter on [Flathub](https://github.com/flathub/flathub/wiki/App-Submission).

**Comment by ashahir05** [comment]
This could be useful since many users tend to want to use flatpaks instead of snaps.

**Comment by 12people** [comment]
I personally refuse to use Snap, as its backend is proprietary. Many distros have opted against supporting Snaps for the same reason, so even for developers who want to support as many distros as possible, Snaps are not a good choice.

I would definitely love Flatpak support.

**Comment by gspencergoog** [comment]
If someone who is familiar with Flatpaks and how to create one can describe the steps necessary, that would help us prioritize this.

**Comment by darkdragon-001** [comment]
@barthalion @alexlarsson Can someone of you have a look? This could help a lot of applications becoming available on Flathub!

**Comment by ashahir05** [comment]
@gspencergoog I did write an SDK extension (with some components from another flatpak). It's kind of outdated but the basics did work. However ```flutter doctor``` doesn't work since flatpaks are sandboxed and there is the need for some workarounds.

https://github.com/ashahir05/org.freedesktop.Sdk.Extension.flutter-dart

**Comment by ntorresalberto** [comment]
+1
I'd be great to get more attention to this.

**Comment by kf6gpe** [comment]
cc @csells --- do we have an "owner" for the open-source installer work that's been done?

**Comment by darkdragon-001** [comment]
> If someone who is familiar with Flatpaks and how to create one can describe the steps necessary, that would help us prioritize this.

@gspencergoog There is a really good [documentation](https://docs.flatpak.org/en/latest/) with the steps necessary to create flatpaks. The important part is the json/yaml description (see @ashahir05's [work](https://github.com/ashahir05/org.freedesktop.Sdk.Extension.flutter-dart/blob/master/org.freedesktop.Sdk.Extension.flutter-dart.json) for a start).

For flathub builds, there is a [contributing guide](https://github.com/flathub/flathub/blob/master/CONTRIBUTING.md). When you create the PR, the maintainers like @barthalion @alexlarsson will help you to bring it in good shape before publishing.

**Comment by jjardon** [comment]
> If someone who is familiar with Flatpaks and how to create one can describe the steps necessary, that would help us prioritize this.

In case it helps, this is an example of what is needed for package a flutter app: https://github.com/flathub/im.fluffychat.Fluffychat/blob/master/im.fluffychat.Fluffychat.json

**Comment by gauthamkrishna9991** [comment]
Can we get a timeline for when this could be provided? I've tried contacting Flathub maintainers, but it seems they might not be the ones who needs to work on this, but rather the maintainers of the software themselves. I feel like this is a very important extension to have, especially when Android Studio already has a flatpak-based app, and this can be added as an extension like so:
![Screenshot from 2021-12-22 12-00-45](https://user-images.githubusercontent.com/32835351/147046356-948c1ae8-4b21-4457-86b7-f12876537db1.png)

**Comment by hotlittlewhitedog** [comment]
Hi @LeandroStanger @gauthamkrishna9991 
Do you have any news of it?  
I would like to publish my flutter app with flatpak and it's actually a real mystery how to build a flutter app with flatpak! 
For info, my app is already on iOS, Mac, Android, Linux (snap).  
Unfortunately the manifest of Fluffychat is also obscure!  

https://discourse.flathub.org/t/flutter-flatpak-request/572

**Comment by lig** [comment]
Personally, I believe that supporting Flatpak is the only way a multiplatform app could benefit from Linux support in Flutter.
Having a pack of more popular platforms on hands it would almost impossible to expect that a developer (a person or a company) decides to support a wide variety of Linux distributions. Out of the box Flatpak support in Flutter would give a developer almost a complete coverage of Linux variations available on the market.

**Comment by Merrit** [comment]
After much headache and frustration figuring out how Flatpak & Flathub work I've gotten everything working (and published a Flutter app on Flathub).

I made a simple example of how to do this based on the counter template app:

https://github.com/Merrit/flutter_flatpak_example

**Comment by gauthamkrishna9991** [comment]
> Hi @LeandroStanger @gauthamkrishna9991 
> Do you have any news of it?  
> I would like to publish my flutter app with flatpak and it's actually a real mystery how to build a flutter app with flatpak! 
> For info, my app is already on iOS, Mac, Android, Linux (snap).  
> https://snapcraft.io/bible-multi-the-life
> 
> Unfortunately the manifest of Fluffychat is also obscure!  
> 
> https://discourse.flathub.org/t/ideas-on-growing-the-flathub-community-in-2022/2053/6
> https://discourse.flathub.org/t/flutter-flatpak-request/572

Nothing yet. I would like to check it out, though I haven't worked on Flatpak Packaging. I've checked out the sources for Steam to understanding how it is that they work, but there seems to need a bit more documentation on how to tackle this exact thing.

**Comment by prateekmedia** [comment]
@Merrit One thing still missing from this example is that flatpak supports build from source and also favors that, but your example only used precompiled linux build to make flatpak app.

Disadvantage of this is that flatpak supports arm linux devices, but this app was only build for x86_64 machines.

**Comment by Merrit** [comment]
> @Merrit One thing still missing from this example is that flatpak supports build from source and also favors that, but your example only used precompiled linux build to make flatpak app.
> 
> Disadvantage of this is that flatpak supports arm linux devices, but this app was only build for x86_64 machines.

In the example is a [github workflow](https://github.com/Merrit/flutter_flatpak_example/blob/main/.github/workflows/flatpak.yml) that builds from source, and packages in flatpak.

Really a flatpak is just a compiled application with the files in the correct directory structure, then wrapped up. So you can compile anything you wish and put the files in the correct locations as shown :)

**Comment by prateekmedia** [comment]
> Really a flatpak is just a compiled application with the files in the correct directory structure, then wrapped up. So you can compile anything you wish and put the files in the correct locations as shown :)

Yes but as I said earlier you are only building for x86_64 platforms and not for arm devices like rasberry pi. Also github actions currently doesn't support building for arm devices, but flatpak does.

So that's why I was telling you about the build from source aspect of flutter.

Also flathub maintainers prefer applications which are build from sources instead of precompiled ones.

**Comment by Merrit** [comment]
> Yes but as I said earlier you are only building for x86_64 platforms and not for arm devices like rasberry pi. Also github actions currently doesn't support building for arm devices, but flatpak does.
> 
> So that's why I was telling you about the build from source aspect of flutter.
> 
> Also flathub maintainers prefer applications which are build from sources instead of precompiled ones.

Yes, the flatpak builder does support its' own compiling, however I found the system frustrating and the documentation frequently very lacking - so I opted to simply use it to assemble the bundle.

As it is it works very well. If someone wants to figure out how to get their builder to compile Flutter apps I'd be happy for the contribution to the example.

**Comment by fabiancrx** [comment]
Would this be better addressed via a plugin as the distribution of the linux app should be up to the developer, and in linux we have many flavors. 
I could imagine a `flatpak` plugin which would work like [flutter_launcher_icons](https://pub.dev/packages/flutter_launcher_icons) , and generate most of the boilerplate, or at least a minimal template and defer to the flatpak docs for further customization.

**Comment by flexagoon** [comment]
@fabiancrx this issue is not about packaging Flutter apps as Flatpak. It is about packagint the official Flutter compiler as a Flatpak.

**Comment by eshimoniak** [comment]
[I've started working on a Flatpak for the Flutter SDK](https://github.com/eshimoniak/flutter-flatpak/). So far the flatpak builds successfully, but there's some runtime errors that seem to be caused by the sandbox.

**Comment by gauthamkrishna9991** [comment]
> > Really a flatpak is just a compiled application with the files in the correct directory structure, then wrapped up. So you can compile anything you wish and put the files in the correct locations as shown :)
> 
> Yes but as I said earlier you are only building for x86_64 platforms and not for arm devices like rasberry pi. Also github actions currently doesn't support building for arm devices, but flatpak does.
> 
> So that's why I was telling you about the build from source aspect of flutter.
> 
> Also flathub maintainers prefer applications which are build from sources instead of precompiled ones.

@prateekmedia The design of Flatpak means that you can run tools in a flatpak itself using QEMU's user-mode emulation (useful for emulating a few `x86_64` apps on `ARM`, though not everything works). Since flatpak packages all dependencies together, and doesn't need anything other than the runtime, you can emulate apps directly on x86_64 system (by installing `qemu-user-static-aarch64` on x86_64 Fedora at least). I've tried this before, testing [Gedit build on `ARM64`](https://www.reddit.com/r/linux/comments/n6x63p/til_that_flatpak_apps_can_emulate_nonnative_apps/) and it worked fairly well, running ARM64 Flatpak apps over x86_64 CPU (though it's kind of iffy, as non-GNOME apps don't as well work yet).

You could conceivably install an SDK for ARM64, and then build your app inside the environment, or you can use ARM64 build of GNOME Builder to work out the project. Performance though would still be lower than if you'll have a cross-compiler (as you're emulating the compilation itself, and that cross-compilers run directly on your system and targeting another).

Again, Flathub prefers that you put the sources in anyway though.

**Comment by hfiguiere** [comment]
One thing nobody mentions is that the current flutter SDK requires to live in a writable location to just work[^1]. Which break many UNIX paradigm, from regular distro (things installed in `/usr`) to Flatpak where the whole SDK and Runtime are read-only.

Without that there is no chance to be able to have a proper SDK.

[^1]: if there is a workaround, nobody knowns about it

**Comment by meutraa** [comment]
> 

The nix people have a patch applied to the sdk because of this.

https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/development/compilers/flutter/patches/move-cache.patch

**Comment by meutraa** [comment]
> One thing nobody mentions is that the current flutter SDK requires to live in a writable location to just work[1](#user-content-fn-1-76ff93e47c9b5b115368ff1834e2ac29). Which break many UNIX paradigm, from regular distro (things installed in `/usr`) to Flatpak where the whole SDK and Runtime are read-only.
> 
> Without that there is no chance to be able to have a proper SDK.
> 
> ## Footnotes
> 1. if there is a workaround, nobody knowns about it [leftwards_arrow_with_hook](#user-content-fnref-1-76ff93e47c9b5b115368ff1834e2ac29)

The nix people have a patch applied to the sdk because of this.

https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/development/compilers/flutter/patches/move-cache.patch

**Comment by KurtB03** [comment]
Would someone would fill me in how flatpaks support for lsp and profiler api is.
And how it would be used in CD/CI pipelines.

I do see some hurdles for the full flutter experience

**Comment by kate-shine** [comment]
> > If someone who is familiar with Flatpaks and how to create one can describe the steps necessary, that would help us prioritize this.
> 
> In case it helps, this is an example of what is needed for package a flutter app: https://github.com/flathub/im.fluffychat.Fluffychat/blob/master/im.fluffychat.Fluffychat.json

As much as it flutters me that Fluffy is shared as an example, it's a hacky one. Copying artifacts that are built in our pipeline over kinda works, but it's error prone and forces us to maintain a separate CI runner just for arm builds. Ideally, the whole flutter app should be built in the Flatpak build itself. Thanks to anyone who is putting their work into making it possible :)

**Comment by CharDSonDev** [comment]
@gasinvein @bilelmoussaoui  Can someone of you have a look?

**Comment by gasinvein** [comment]
@CharDSonDev As @hfiguiere said earlier in this thread, flatpak'ing Flutter is not possible as long as it requires its installation to reside in a writable location.

**Comment by 12people** [comment]
Since Flutter does not have a GUI, it doesn't seem Flatpak is a good fit for packaging it.

[Toolbx](https://containertoolbx.org/), which is used for containerizing CLI environments and recommended for use in immutable OS's like Fedora Silverblue, seems like a much better fit. It seems that someone already created a Toolbx image for Flutter: [flutter-toolbx](https://github.com/JamesBelchamber/flutter-toolbx). I haven't tried it yet, but perhaps that might solve the problems that people here are dealing with.

Of course, this is only a part of the puzzle.

Ideally, going forward, Flatpak-based IDEs would integrate Toolbx and provide GUI options to install common Toolbx images.

And there's also the question of documentation for building Flatpaks. I created a separate issue for that over on the Flutter Website repo, you can find it here: https://github.com/flutter/website/issues/7641

**Comment by bilelmoussaoui** [comment]
> Since Flutter does not have a GUI, it doesn't seem Flatpak is a good fit for packaging it.

That is completely wrong. Flatpak is used to ship SDKs so that an application built on top of Flutter can use it to distribute it application as a Flatpak.

**Comment by 12people** [comment]
A separate SDK is not needed to **ship** Flutter apps, as Flutter apps are shipped as binaries. There are a [number of](https://flathub.org/apps/details/im.fluffychat.Fluffychat) [Flutter](https://flathub.org/apps/details/com.enjoyingfoss.Parlera) [apps](https://flathub.org/apps/details/io.github.prateekmedia.appimagepool) [distributed](https://flathub.org/apps/details/com.github.KRTirtho.Spotube) [this way](https://flathub.org/apps/details/io.github.prateekmedia.pstube) as Flatpaks already. @Merrit posted a repo with an [example of a Flutter-based Flatpak project](https://github.com/Merrit/flutter_flatpak_example).

It would be possible to build Flatpak apps on-device, I guess, but then Flutter wouldn't be shipped as a Flatpak itself, but rather as an [SDK for Flatpak](https://docs.flatpak.org/en/latest/building-introduction.html#software-development-kits-sdks) — big difference. If that's what this issue is about, it should be renamed to make that clear, or a separate issue should be created.

**Comment by meutraa** [comment]
> A separate SDK is not needed to **ship** Flutter apps, as Flutter apps are shipped as binaries. There are a [number of](https://flathub.org/apps/details/im.fluffychat.Fluffychat) [Flutter](https://flathub.org/apps/details/com.enjoyingfoss.Parlera) [apps](https://flathub.org/apps/details/io.github.prateekmedia.appimagepool) [distributed](https://flathub.org/apps/details/com.github.KRTirtho.Spotube) [this way](https://flathub.org/apps/details/io.github.prateekmedia.pstube) as Flatpaks already. @Merrit posted a repo with an [example of a Flutter-based Flatpak project](https://github.com/Merrit/flutter_flatpak_example).
> 
> It would be possible to build Flatpak apps on-device, I guess, but then Flutter wouldn't be shipped as a Flatpak itself, but rather as an [SDK for Flatpak](https://docs.flatpak.org/en/latest/building-introduction.html#software-development-kits-sdks) — big difference. If that's what this issue is about, it should be renamed to make that clear, or a separate issue should be created.

Flutter **is** an SDK. I don't see any ambiguity in the title. I don't know why this issue needs any discussion of applications built using flutter. This is a packaging request for flutter, for developers.

**Comment by meutraa** [comment]
> @CharDSonDev As @hfiguiere said earlier in this thread, flatpak'ing Flutter is not possible as long as it requires its installation to reside in a writable location.

https://github.com/flutter/flutter/issues/91318 is the relevant issue for this.

**Comment by 12people** [comment]
> Flutter **is** an SDK. I don't see any ambiguity in the title. I don't know why this issue needs any discussion of applications built using flutter. This is a packaging request for flutter, for developers.

Flutter is an SDK, but making a **Flatpak** out of it is something completely, entirely different than making a **shared Flatpak runtime** out of it.

For the former case, there are some SDKs that go the route of being bundled with an IDE — e.g. [Processing](https://flathub.org/apps/details/org.processing.processingide), and even [Android Studio](https://flathub.org/apps/details/com.google.AndroidStudio) offers downloading the Android SDK, which is then available just to that Android Studio Flatpak.

[Runtimes](https://docs.flatpak.org/en/latest/basic-concepts.html?highlight=runtime#runtimes), on the other hand, are shared across Flatpaks, but only used to build the apps themselves. They're primarily intended for the compilation of specific Flatpak apps, not as a way to add libraries to IDEs. That's why there is no Android SDK Flatpak runtime.

Creating a Flatpak runtime would allow Flatpak apps to be built client-side, rather than shipping a binary as a Flatpak. I suppose it's a nice-to-have, though I don't see much of a need for it. Is that what this issue is about?

For IDE use and integration, however, Flatpak is not an appropriate tool; Toolbx is. And this doesn't go just for Flutter, but for all SDKs for development use.

**Comment by darkdragon-001** [comment]
Maybe an SDK Extension is the right thing? Something like [OpenJDK](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk).

This can then be used to easily build and package flutter apps and distribute them in binary form.

**Comment by 12people** [comment]
> Maybe an SDK Extension is the right thing? Something like [OpenJDK](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk).
> 
> This can then be used to easily build and package flutter apps and distribute them in binary form.

An SDK extension would make it possible to build the app on the fly on a user's device. (Just like JRE runs on a user's device so that an app can be run on top of it.)

No extension is needed for distribution of Flutter apps in binary form — that is possible already with the current Flutter runtimes, as evidenced by the Flutter-based Flatpaks that exist already. I linked to some of them in my [post above](https://github.com/flutter/flutter/issues/41737#issuecomment-1273083018).

That said, I'm in favor of an SDK extension like this, but I didn't think that's what this issue was about and it certainly wasn't worded as such. There are people in this thread who don't know how to build Flutter apps for Flatpak, and I thought this thread was for them. An SDK extension wouldn't solve that problem, good documentation would, as it's already possible to build Flatpaked Flutter apps.

If the conversation was about an SDK extension all along, then this issue should really be renamed.

**Comment by darkdragon-001** [comment]
But if one wants to actually build the application from source via Flatpak manifest instead of downloading an archive with binary files (e.g. for easier multi-arch distribution), it makes life a lot easier if one can just specify the SDK extension, download the application source and then run some script from the SDK extension to build and bundle the result.

**Comment by 12people** [comment]
> But if one wants to actually build the application from source via Flatpak manifest instead of downloading an archive with binary files (e.g. for easier multi-arch distribution), it makes life a lot easier if one can just specify the SDK extension, download the application source and then run some script from the SDK extension to build and bundle the result.

Agreed. It's the "shipping binaries" bit that you wrote that threw me off.

Compiling on a user's device instead of shipping binaries will have both upsides and downsides. One upside would be that supporting multiple architectures (ARM + x86_64) would no longer be a hassle, as Flutter apps cannot currently be cross-compiled. Some downsides would be longer install times (that would now include building the app itself) as well as extra space taken up on the user's device by Flutter libraries.

As I said, I agree with having an SDK extension. However, that's not at all what I understood by the issue description. The initial post reads "Please, consider supporting Flatpak and publishing flutter on Flathub," but Flutter already supports Flatpak in the form of binaries. That's just not documented. And using Flutter inside of Flatpak IDEs needs to be done using Toolbx or similar, Flatpak is not the right tool for that.

**Comment by darkdragon-001** [comment]
I agree.

Maybe what could be done on the Flutter side to ease it, is generating icon, `.desktop` file and `metainfo.xml` and maybe even the Flatpak manifest directly with the SDK such that Flatpak packaging just needs to unpack the tgz without additional files.

**Comment by hotlittlewhitedog** [comment]
During this time, Ubuntu Flutter Community does this:
https://www.omgubuntu.co.uk/2022/08/new-unofficial-ubuntu-software-app-flutter
👍👍👍

**Comment by hotlittlewhitedog** [comment]
I found this:

https://medium.com/kinandcartacreated/setup-a-flutter-ci-cd-with-gitlab-ci-part-1-8d40b7a3b3dc

**Comment by 12people** [comment]
There was [an effort to create an official SDK extension for this](https://github.com/flathub/flathub/pull/3777), but it got stuck on these two issues:

- #91318
- #22429

Bring attention to those two issues, as it doesn't look like this issue can be solved before those two are.

**Comment by iapicca** [comment]
> There was [an effort to create an official SDK extension for this](https://github.com/flathub/flathub/pull/3777), but it got stuck on these two issues:
> 
> * [[flutter_tools] Refactor Flutter SDK to support installation in read-only directory #91318](https://github.com/flutter/flutter/issues/91318)
> * [Offline mode needs to be extended to all flutter commands #22429](https://github.com/flutter/flutter/issues/22429)
> 
> Bring attention to those two issues, as it doesn't look like this issue can be solved before those two are.

[flatpack has reached 1M active users](https://docs.flathub.org/blog/over-one-million-active-users-and-growing/)maybe is time to re-evaluate this issue

@TahaTesser shouldn't this be at least `crowd`?

**Comment by mbnoimi** [comment]
> This could be useful since many users tend to want to use flatpaks instead of snaps.

Personally I dropped snap forever. Even most popular Linux distros dropped it (ex. Linux Mint)

**Comment by flexagoon** [comment]
By the way, while a Flatpak version of Flutter would be great, people who just want to get Flutter working on Fedora Silverblue (or just install flutter without cluttering your system) can use a Distrobox image that I made for myself:

https://github.com/flexagoon/flutterbox

It installs flutter with all of the dependencies in a container and allows you to run flutter and dart from the host system

**Comment by Karlosbubi** [comment]
I am still missing how/since when flatpak supports something like the exports you did from distrobox. And I have considerable doubts regarding the usefulness of packaging SDKs without exposing them at least as much.

**Comment by TDuffinNTU** [comment]
Poking this again. Keeping Flutter tied to snap is a deathknell for the SDK on Linux, especially as it becomes increasingly difficult to install snap packages outside of the Ubuntu ecosystem; it's particularly annoying to have to work around installing a whole repository just for two packages (android studio and Flutter), especially if that system is immutable.

Flutter works really well otherwise, but I do think it's time that Flutter as a project considers moving away from snap to a truly platform-agnostic alternative; whether that's flatpak, appimage, or just building from source altogether.

**Comment by Karlosbubi** [comment]
Snap is absolutely fine for flutter. As much as enthusiasts hate them, Ubuntu, RHEL and SUSE are virtually the only distros used by enterprises and all work well with snap. It won't hurt flutter at all if not every hobbyist builds their 20 minute 3 button apps in it.
Regardless i do support the sentiment, to publish flutter in binary packages (.deb & .rpm) or via the JetBrains Toolbox, but it's just convenience

**Comment by bilelmoussaoui** [comment]
> Snap is absolutely fine for flutter. As much as enthusiasts hate them, Ubuntu, RHEL and SUSE are virtually the only distros used by enterprises and all work well with snap. It won't hurt flutter at all if not every hobbyist builds their 20 minute 3 button apps in it. Regardless i do support the sentiment, to publish flutter in binary packages (.deb & .rpm) or via the JetBrains Toolbox, but it's just convenience

No one talked about Jetbrains Toolbox btw...
